Fblock addition

Emyrk opened the pull request.

Emyrk committed:
Add Redem Conditional Datastructure

The Redeem Conditional Datastructure is underlying datastructure
to a factoid address. It is provided in factoid transactions
corrosponding 1 to 1 with a fct input. Currently
only rcd type 1 is supported.

Emyrk committed:
Add Factoid Transactions

Included factoid transactions marshal/unmarshal. Not included
is the composing of factoid transactions.

Emyrk committed:
Made transaction stream decoding

Transaction decode will now return the number of bytes read
to be used when decoding serial transactions. Like in an fblock

Emyrk committed:
Add FBlock functionality and handling

FBlock marshal/unmarshal + computation of keymr and ledgermr.

Emyrk committed:
Add missing json tags

Emyrk committed:
Add unit test to excersise varint decode panic

Varinf.Decode can panic from invalid data. This should
never happen, it should return the error condition
response.

Emyrk committed:
Add FactoidTransaction.Valid()

Factoid transactions can now be self checked for validity
of their inputs and signatures.

Emyrk committed:
Add RCD validate function

RCDs can now validate signature blocks for a given message.

Emyrk committed:
Added additional unit tests for block unmarshaling

Ensuring fblock unmarshal does not panic with unit tests

Emyrk committed:
Add Fblock.Get support for height field

Added a unit test to "query" factomd with a fixed
response to test the "Get"

Emyrk committed:
Add utility to mock factomd query responses

Emyrk commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+	// It is the encoded bytes into the human readable addresses
+	Address *Bytes32 `json:"address"`
+
+	// TODO: Do you want to also put the human readable address when we unmarshal?
```
See comment

Emyrk committed:
Handle invalid varint reads

Emyrk committed:
Add unit tests to ensure ispopulated can detect the empty struct

AdamSLevy commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+// Transaction efficiently finds and returns the *FactoidTransaction in fb.Transactions
+// for the given txid, if it exists. Otherwise, Transaction returns nil.
+func (fb FBlock) Transaction(txid Bytes32) *FactoidTransaction {
+	ei := sort.Search(len(fb.Transactions), func(i int) bool {
```
Doesn't this require that fb.Transactions be sorted by ChainID? I thought these were in chronological order?

Emyrk replied:
Oh I totally messed that up. I thought I had done a sort first, so this function is incorrect.

AdamSLevy replied:
If looking up a specific txid isn't needed at this time I would omit this. This requires maintaining a sorted slice. and if this is needed by a client later, then they can be better off with a map going from txids to an index in the Transactions map.

Emyrk replied:
Removed

AdamSLevy commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+
+	expansionSize := varintf.Encode(uint64(fb.Header.ExpansionSize))
+	data := make([]byte, FBlockMinHeaderLen+len(expansionSize)+len(fb.Header.ExpansionBytes))
+	i := 0
```
when simply declaring a variable to zero, simply use the `var i int` syntax by convention. Use short declaration syntax when declaring a variable and initializing it to a nonzero value

AdamSLevy commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+	return data, nil
+}
+
+func (fb *FBlock) MarshalBinaryHeader() ([]byte, error) {
```
Because of my change to how and when I'm computing KeyMRs and ChainIDs, I no longer needed to have the MarshalBinaryHeader function to be separate from MarshalBinary. I'd like to see the same changes reflected here eventually. Please just move this into MarshalBinary. Please also then use a single allocation for data if the exact full size can be known.

Emyrk replied:
The exact size of the header is known, but not the body. I'd like to keep this separated as I use the MarshalBinaryHeader() in the KeyMr code, since the header is a branch and the body is the other branch. So this function is used outside MarshalBinary

Emyrk replied:
I see what you mean for the changes.
I can move this in, I just like the code separation.

AdamSLevy commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+	}
+
+	// Transactions + minute markers are included
+	leaves := make([][]byte, len(fb.Transactions)+len(fb.endOfPeriod))
```
In my most recent changes, I validate the bodyMR in the unmarshal func. When doing this I avoid additional allocations by building the body elements by takings slices of the incoming data as I parse it. When done I pass that to ComputeEBlockBodyMR or ComputeDBlockBodyMR, which are slightly different. You may be able to reuse those.

AdamSLevy commented on fblock_test.go:
```diff
@@ -0,0 +1,138 @@
+	}
+
+	// To ensure there is not panics and all errors are caught
+	t.Run("UnmarshalBinary/Random", func(t *testing.T) {
```
This is a really great idea for testing. I should do this in my tests more.

AdamSLevy commented on rcd.go:
```diff
@@ -0,0 +1,144 @@
+	// Address returns the sha256(rcd) that is the factoid address
+	Address() (*Bytes32, error)
+
+	Validate(msg Bytes, signature Bytes) bool
```
Keeping interfaces really small is a good idea. This seems like the one most important function that we could have an interface for.

AdamSLevy commented on varintf/varintf_test.go:
```diff
@@ -95,6 +96,34 @@ func TestFactomSpecExamples(t *testing.T) {
 }
 
+// TestVarInt_Panic ensures the varint decoding will never panic from invalid data
+func TestVarInt_Panic(t *testing.T) {
```
I updated varintf to avoid panics. Haven't tested it against this yet.

Emyrk replied:
Test passes now.

AdamSLevy commented on varintf/varintf_test.go:
```diff
@@ -95,6 +96,34 @@ func TestFactomSpecExamples(t *testing.T) {
+	t.Run("test error handling (no panic)", func(t *testing.T) {
+		for i := 0; i < 1000; i++ {
+			d := make([]byte, rand.Intn(20))
+			rand.Read(d)
```
I believe the seed should be explicitly set somewhere to a constant value so that "random" tests are reproducible. Or is the seed always the same if not otherwise initialized?

Emyrk replied:
The seed is the same each time you run unless someone explicitly changes the seed. But the seed is global, so there is some value is initializing it upfront. Since even adding a unit test, or changing the execution order would change the values.

I would argue that those subtle changes though are fine. As long as the tests are reproducible, given you know the test conditions (`go test -run TestXYZ` vs `go test ./...`), I think that is sufficient.

AdamSLevy commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+
+type FactoidTransactionIOs []FactoidTransactionIO
+
+type FactoidTransactionIO struct {
```
I had a big thing written about how this should be done with a map. But I took some time to think it over and, hey, this works, and it preserves exact order of inputs and outputs in a simple way. It would be really nice if inputs and outputs had to be ordered but that's not the case so oh well. 

For users later wishing to construct transactions, catching duplicate addresses is important. But for fatd or pegnet, not so much. But the possibility of such should be taken into account. When parsing all amounts I still would recommend converting this to a map. but doing so for every transaction all the time would be more memory intensive.

AdamSLevy commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+func (f FactoidTransaction) IsPopulated() bool {
+	return f.FCTInputs != nil && // Although a coinbase has 0 inputs, this array should not be nil
+		f.Signatures != nil &&
+		f.Timestamp != time.Time{}
```
time.Time has an IsZero function for this purpose

Emyrk replied:
TIL

AdamSLevy commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+	Amount uint64 `json:"amount"`
+	// Address can be an SHA256d(RCD) for FCT in/out, or a public key for EC out.
+	// It is the encoded bytes into the human readable addresses
+	Address *Bytes32 `json:"address"`
```
Feel free to just embed FAAddress directly, or just name this FAAddress if embedding is not what you want here. But don't call it Address as that is the name of an interface in this package which is confusing when it represents a different type.

Emyrk replied:
Sure, unfortunately, this is used for EC and FA addresses. So it's either an EC of FA address. I did not want to make a different IO type. I'm trying to think of a good name, but don't have anything atm.

AdamSLevy replied:
hmm. ok. I'm going to think on this. it may be better to have specific types that are redundant. or perhaps this is a place for the Address interface

Emyrk replied:
Let me know what you decide here. Atm I will keep it as a Bytes32

AdamSLevy commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+		return err
+	}
+
+	txid, err := f.ComputeTransactionID()
```
Move this to UnmarshalBinary. Populate the txid if nil, otherwise validate it. Populate it prior to Unmarshaling to then ensure that it is validated.

AdamSLevy commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+		return nil, fmt.Errorf("not populated")
+	}
+
+	// It's very difficult to know the size before marshaling, as each in/out has a varint
```
Check out my new varintf package API. I added a BufLen() function which determines the exact length of a varintf encoding for a given uint64. I also added a Put function to allow the user to control allocations.

Emyrk replied:
Each input/output in the body has a varint. To compute the size up front is not trivial.

AdamSLevy commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+
+// MarshalLedgerBinary marshals the transaction ledger fields to their binary representation.
+// This excludes the signature blocks
+func (f *FactoidTransaction) MarshalLedgerBinary() ([]byte, error) {
```
Since computing hashes, keymrs, etc, should now be moved to the unmarshalBinary function where it can be done initially more efficiently on incoming data, the marshal functions do not need to be split up to accommodate the computation of keymrs and hashes after the fact. 

I'm not sure it adds all that much clarity to have these things broken up anymore and since the other marshaling functions now likely won't be used in any other context, they won't be reused anywhere. The other advantage to putting them in a single function is that it makes the allocations very explicit and clear.

Emyrk replied:
Currently, it is used in the Valid() function. So if you build a tx with this library, you would need this function to validate the signatures.

I did however remove the use of this in the Decode to compute the txid, so the decode follows the unmarshal idea of not reallocating bytes.

Emyrk replied:
We decided to keep it because users can build txs, and this is needed to verify signatures.

AdamSLevy commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+func (f *FactoidTransaction) Decode(data []byte) (i int, err error) {
+	// Because the length of an FactoidTransaction is hard to define up front, we will catch
+	// any sort of out of bound errors in a recover
+	defer func() {
```
I do not like this. It will hide bugs and make the difficult to uncover. And doing it one place encourages it to be used again in others. 

I believe the code can be made more precise and explicit by doing better bounds checking as you go. But I'd like to hear about where it is difficult to achieve this.

Emyrk replied:
I'll attempt to add better error checking, but there was a lot of variable length things that we need to check. I will come up with the complications and work on unit tests to break them.

AdamSLevy replied:
the new varintf API may help some

Emyrk replied:
I removed all panic to recover defers and added enough checks to pass the tests. I don't know if some malformed data can still cause a panic, but it's much more robust now.

AdamSLevy commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+	return err
+}
+
+func (f *FactoidTransaction) DecodeHeader(data []byte) (int, error) {
```
Lets move this directly into Decode. This does not abstract something that gets reused elsewhere, so it forces the reader to visit another section of code for not a lot of good reason. Do this inline in Decode.

AdamSLevy commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+// calling this function. It will return how many bytes it read and return an error.
+func (ios FactoidTransactionIOs) Decode(data []byte) (int, error) {
+	i := 0
+	for c := 0; c < len(ios); c++ {
```
use a range for loop here as it improves readability.

AdamSLevy commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+func (ios FactoidTransactionIOs) Decode(data []byte) (int, error) {
+	i := 0
+	for c := 0; c < len(ios); c++ {
+		read, err := ios[c].Decode(data[i:])
```
I wonder if all this Decode nesting couldn't be reduced. but we'll leave it for now. However it might be a good idea to unexport a lot of these Decode functions to simplify the exported API. Outside of this package I doubt anyone will need to decode just a FactoidTransactionIOs. That slice type may also not need to be exported either. Anything that won't be explicitly needed outside the package should not be exported to keep the API small.

Emyrk replied:
Those are good points. I would like to keep the Decode on the tx exported. For myself as a developer, I dislike when a libary thinks I do not ever need something. But the nested decodes we could probably hide, like the slice implementations.

Emyrk replied:
Although the API exported is exported on the structures themselves, not the top level. So it's not a lot of overhead on the user to keep it in. But I can live with hiding a lot of these decodes.

AdamSLevy replied:
Any exposed API is something that we later may be locked into. That's why I want to keep it small.

I think it's important to keep in mind the specific use case of these data structures to the users of this package: primarily factom clients, specifically fatd. If it isn't clearly needed for those two, I'd like to keep it unexported. It is easier to add these things later than remove them.

Emyrk replied:
I hid the FactoidTransactionsIO data type that I use to not have to write the same for loop each time. Let me know which functions you'd like to hide.

AdamSLevy commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+// the set of factoid transactions ios. The set length should be preset before
+// calling this function. It will return how many bytes it read and return an error.
+func (ios FactoidTransactionIOs) Decode(data []byte) (int, error) {
+	i := 0
```
var i int

AdamSLevy commented on rcd.go:
```diff
@@ -0,0 +1,144 @@
+	//		until that is fixed, the only way to proctect the call is
+	// 		to panic recover
+	defer func() {
+		// TODO: Remove this recover if the varintf ever gets a protected call
```
varintf is updated.

AdamSLevy reviewed (commented):
Sorry for the delay in getting this back to you. This is really great stuff and I went back to the package and updated a number of things that I either learned from your code or that I had to rethink as a result of this review.

When reading this review, I recommend starting in the order I did it in:
- rcd.go
- transaction.go
- fblock.go

Emyrk commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+
+	Header FBlockHeader `json:"header"`
+
+	Transactions []*FactoidTransaction `json:"transactions"`
```
What is your argument for not using them?

I noticed you had not (noticed today when I am actually using the repo). It makes things like this work. 
```
for _, tx := range fblock.Transactions {
     tx.Get(c)
}
```
Without you have to do:
```
for i, _ := range fblock.Transactions {
     fblock.Transactions[i].Get(c)
}
```

I just don't see the benefit of not using a pointer in the slice, except for the potential bug.

AdamSLevy replied:
yes and that is how it should be. My coding mantra, stolen from Bill Kenedy, is "don't make things easy to do, make things easy to understand". Using a pointer hides the fact that you are modifying the underlying slice in that value semantic for loop. 

Seeing this:  (don't use that second underscore in this case btw)
```
for i := range fblock.Transactions {
     fblock.Transactions[i].Get(c)
}
```
makes it very clear that my intent is to index into the slice I'm ranging over. If I need to reuse that same tx a lot I do this at the top:
```
for i := range fblock.Transactions {
     tx := &fblock.Transactions[i]
    tx.Get(c)
 ...
}
```

Seeing this: 
```
for _, tx := range fblock.Transactions {
...
}
```
makes it really clear that I'm only interested in the values within the slice, and not in modifying the slice itself.

These are features of golang I want to be properly leveraged for clarity. Using a slice of pointers obfuscates that clarity that the syntax is allowing us to take advantage. So it will deceive people unless they go back and read the slice definition, which they might have skipped over because it is not obviously that there is anything extra to pay attention to.

Emyrk replied:
I can live with it. When I was first using this library, I fell for that pitfall, and that is when I realized you didn't use pointers for the eblocks -> entries. I probably would have copied that if I had noticed that before I made the transactions here.

Emyrk commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+
+	if fb.KeyMR != nil {
+		orig := fb.KeyMR
+		defer func() { // Ensure we got the block we asked for
```
It is only supposed to be run in the if statement. I wanted to ensure the block you retrieved was the same block you requested. Your other comment says to check it in the unmarshal binary, so I can move it to there.

Emyrk commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+// [PrevLedgerKeyMR (Bytes32)] +
+// [Exchange Rate (8 bytes)] +
+// [DB Height (4 bytes)] +
+// [Header Expansion size (Bytes)] +
```
I do not have the answer to that offhand. I'd have to bring that question up with Brian and Paul.

Emyrk replied:
So just something I learned recently. An expansion thing we might add is a balancehash for leaders to check their balances against each other. This would be an optional thing for nodes to check against, and mainly for the authority set.

So instead of making a hard fork, we can use the expansion space for a feature like this, without impacting any running nodes.

Emyrk commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+func (fb *FBlock) UnmarshalBinary(data []byte) (err error) {
+	// Because the length of an FBlock is hard to define up front, we will catch
+	// any sort of out of bound errors in a recover
+	defer func() {
```
I agree I do not like using a recover(), but there is a lot of unknown lengths here. An fblock contains all txs in their entirety. Each tx has a variable number of inputs/outputs and varints for the amounts. I will probably go through and try to catch more, but there is so many edge cases, not just out of bounds panics.

E.g: Someone could send you data saying there is maxint number of inputs, and you'd try to allocate a factoid tx slice outside the limits of your system memory. 

I do want to add more protections, but this was my catch all, as I'd need to think of all the possible ways to crash this. And it's more than the other structures.

AdamSLevy replied:
If we run out of memory there is nothing we can do and that is the exact situation where an unrecovered panic is appropriate. 

This recover hides programming or deeper errors and reduces them to a single error with no context.

Emyrk replied:
I am saying if they try to unmarshal a blob of data is that totally malformed, currently it will panic.

As a developer, I assume if I pass a bad fblock data blob into this function, it will give me an error if there is a problem with the data. I don't expect panics from libraries, as panics are not like exceptions.

The OOM panic is because the binary data is wrong, and it says to allocated some massive number of transactions, which obviously fails. We can put safety checks at places, but we'd be choosing arbitrary magic numbers for the maximum number of txs in a block for example

Emyrk replied:
I did as much as I could to catch things and dropped the defer.

Emyrk commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+	fb.Header.Height = binary.BigEndian.Uint32(data[i : i+4])
+	i += 4
+
+	expansionsize, read := binary.Varint(data[i:])
```
That varint use was a mistake, good catch. I probably let my IDE choose that varint implementation. The idea behind the current unmarshal is that it won't break your code if you do not update.

In the admin block for example, we wrote a marshal function that handles all future admin block types. So when you decode an admin block, if you do not understand it, you just pack it into a byte slice and ignore it.

For this section, if someone if using your library and following fblocks, if the expansion stuff is used, it won't break any clients. They will still be able to read factoid txs, because the marshal will just pack that away, and ignore it.

Emyrk commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+	i += read
+	fb.Header.ExpansionBytes = make([]byte, expansionsize)
+	copy(fb.Header.ExpansionBytes, data[i:i+int(expansionsize)])
+	i += int(expansionsize) // TODO: This should be safe, as the expansion size is never > max int
```
Fair. I can make the int32 assertion, I doubt we will ever reasonable exceed a 4gb fblock.

AdamSLevy replied:
Actually the issue is that the expansion size could theoretically be up to maxUint64. We need to check for a sane value before allocating the ExpansionBytes. Very simply, the expansion bytes cannot be larger than the remaining size of data, less any other data we can be guaranteed to have follow, if we can count it at this point.

Emyrk replied:
I added a check about the length of the data.

Emyrk commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+	fb.Header.BodySize = binary.BigEndian.Uint32(data[i : i+4])
+	i += 4
+
+	// TODO: Body
```
Yes :cry:

Emyrk commented on rcd.go:
```diff
@@ -0,0 +1,144 @@
+// RCD is the underlying structure behind a factoid address. A factoid address is a sha256d(RCD).
+// The most common and basic RCD type, is type 1. That being just a single public key that uses a single
+// 64 byte signature block.
+type RCD interface {
```
I can get on board with this. I really wanted to keep the ability to handle multiple RCDs, but that is a long ways out. The library will fail when a new RCD is released in my current version, so I can live with just assuming everything is RCD 1, and failing otherwise in the current release.

AdamSLevy replied:
Yeah lets keep things simple and specific until new RCD types are defined. then add the concrete implementations and then retro fit code that uses RCDs with the proper and limited interface abstraction.

The other thing to consider is that new RCD types may require a blockheight activation, which add some complexity to very generic implementations where that would not get caught.

Emyrk commented on rcd.go:
```diff
@@ -0,0 +1,144 @@
+
+// RCD1 is the simple rcd of a factoid address with a single public key.
+// RCD1 contains the type and a single 32 byte ed25519 public key
+type RCD1 struct {
```
All good points. I was just using Bytes32, as I know you were using it around the place. The current factom api does return the RCD hex, but there has yet to be an application to expose that in the hex format vs the base58 with the prefix and checksum.

AdamSLevy replied:
Totally. in this case I think it is more specific to declare the array explicitly with the constants defined by the ed25519 package.

Emyrk commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+	"github.com/Factom-Asset-Tokens/factom/varintf"
+)
+
+type FactoidTransaction struct {
```
I was unsure, I felt this was pretty explicit. I am happy either way

AdamSLevy replied:
Might as well leave it this way to be explicit

Emyrk commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+
+	Version uint64 `json:"version"`
+	// Timestamp is accurate to the millisecond
+	Timestamp time.Time `json:"timestamp"`
```
It is the timestamp salt, I like that suggested method.

AdamSLevy replied:
cool. See this comment for more info: https://github.com/Factom-Asset-Tokens/factom/pull/13/files#r329285162

Emyrk commented on transaction.go:
```diff
@@ -0,0 +1,489 @@
+
+type FactoidTransactionSignature struct {
+	// SHA256d(RCD) == FactoidIOAddress for the inputs
+	ReedeemCondition RCD   `json:"rcd"`
```
I am going to assume you mean to change to the type RCD1? That has the validate function on it, which I use in the validation of the tx.

AdamSLevy replied:
yes, I think so. but what I was thinking at the time was that you might not know the RCD type here yet, but since we only have 1 to work with lets just be specific.

Emyrk reviewed (commented):
Some initial comments. Will try to get to implementation soon

AdamSLevy commented on fblock.go:
```diff
@@ -77,7 +77,7 @@ type FBlockHeader struct {
 func (fb FBlock) IsPopulated() bool {
 	return len(fb.Transactions) > 0 && // FBlocks always contain at least a coinbase
 		fb.Header.BodyMR != nil &&
-		fb.Header.PrevKeyMR != nil && // TODO: Why not check the computed keymr?
```
Is populated is primarily used for two purposes: ensuring that MarshalBinary can proceed and allowing Get to exit early so as to avoid making redundant network calls. So what is included here should be chosen with that in mind. KeyMR is not needed to marshal.

Emyrk committed:
Make RCD return FAAddress

Given all RCDs are for factoid addresses, the Address returned
should be the FAAddress type

AdamSLevy commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+	// Transactions + minute markers are included
+	leaves := make([][]byte, len(fb.Transactions)+len(fb.endOfPeriod))
+	period := 0
+	c := 0
```
var period, c int

Emyrk committed:
Correct compile errors following rebase

Rebase had some compile errors. Correcting some zero initialized
variables as well.

AdamSLevy commented on rcd.go:
```diff
@@ -50,7 +50,10 @@ func DecodeRCD(data []byte) (reedemCondition RCD, read int, err error) {
 
-	version, _ := varintf.Decode(data)
+	version, r := varintf.Decode(data)
+	if r == -1 {
```
`if r < 0 {` is slightly more defensive programming.

Emyrk committed:
Adjustments from code review

- Switch variable initialization when 0
- Reduce some comment lengths to be max 80 on line
- Remove panic recover (updated varint)

Emyrk committed:
Fix some compile issues

Emyrk commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+	copy(fb.Header.ExpansionBytes, data[i:i+int(expansionsize)])
+	i += int(expansionsize) // TODO: This should be safe, as the expansion size is never > max int
+
+	fb.Header.TransactionCount = binary.BigEndian.Uint32(data[i : i+4])
```
I'm not sure. Most of the structure existed before me.

Emyrk committed:
Made changes from PR suggestions

- Flatten FBlock header
- Made some RCD changes
- Fix Client_test for jsonrpc version

Emyrk committed:
Remove expansionsize field

Emyrk commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+	// Number of txs in the block
+	TransactionCount uint32 `json:"txcount"`
+	// Number of bytes contained in the body
+	BodySize uint32 `json:"bodysize"`
```
It makes the marshal function create the byte size up front. But if you get the dblock without setting this, then it would fail.

Emyrk commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+)
+
+const (
+	FBlockMinuteMarker = 0x00
```
I'm not sure exactly, but eblocks are a collection of keymrs, all being 32 bytes. Fblocks are a collection of transactions, which are variable in length. So we know a 0x00 between two transactions is a minute marker. In an eblock, a 0x00 between 2 hashes could be the start of the second hash, with a 0x00 at the end. Or is the first 0x00 following 2 hashes... It's not clear. We'd have to add a var length before every hash to indicate a hash vs a minute marker.

Emyrk commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+			}
+			fb.endOfPeriod[period] = c
+			period++ // The next period encountered will be the next minute
+			i += 1
```
I do this alot. Is there any reason why the `i += 1` is bad? We do that `i += ..` for the marshal stuff, I kinda just like it looking the same.

Emyrk committed:
Correct a comment + pass conditional vs eval

Emyrk commented on client_test.go:
```diff
@@ -0,0 +1,55 @@
+	"io/ioutil"
+	"net/http"
+
+	"github.com/AdamSLevy/jsonrpc2"
```
I had to edit this outside GoLand. GoLand would not let me put that v12. The goimports kept removing it.

Emyrk committed:
Correct jsonrpc2 import

Emyrk commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+
+// UnmarshalBinary unmarshals raw directory block data.
+//
+// Header
```
I think I have it like you want it now.

Emyrk committed:
Remove function that was improperly implemented

Emyrk committed:
Correct test type assertion check

Emyrk committed:
Make function receivers the same

Clean up a unit test equal assertion

Emyrk committed:
Remove DecodeHeader and put it inline into Decode

Emyrk committed:
Make the ios helper unexported

Shorten some comment line lengths

Emyrk committed:
Remove counts, calculate them from list lengths

Emyrk commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+	}{fb.Header.Height}
+	result := struct {
+		// We will ignore all the other fields, and just unmarshal from the raw.
+		RawData Bytes `json:"rawdata"`
```
Want me to remove all my json tags?

Emyrk commented on fblock.go:
```diff
@@ -0,0 +1,455 @@
+				return
+			}
+
+			if fb.KeyMR != nil && *orig != *fb.KeyMR {
```
You mean if the keymr is set, then in the unmarshal binary, validate the newly compute keymr matches?

Emyrk replied:
I understand now :+1:

Emyrk committed:
Compute & validate txids on transaction unmarshal

Some code formatting adjustments

Emyrk committed:
Add more checks on the transaction decode

Emyrk committed:
Add better decode guards around the transaction decode

Emyrk committed:
remove recover comments

Emyrk committed:
Add better unmarshal protection

Emyrk committed:
Compute merkle roots in the unmarshal function of the fblock

Emyrk committed:
Remove member functions for keymr computation

Emyrk committed:
Remove extra byte allocation on Decode

Emyrk committed:
Cleanup struct fields and json tags

Emyrk committed:
Flatten MarshalBinaryHeader into MarshalBinary. Add ctx to Get

AdamSLevy committed:
refactor(FBlock,Transaction,RCD): Reduce/simplify the API and memory usage for Unmarshaling

There is now just one RCD type that handles all RCD types dynamically.
